device/iot-device-samples/pnp-device-sample/temperature-controller-device-sample/src/main/java/samples/com/microsoft/azure/sdk/iot/device/PnpConvention.java

```diff
@@ -121,29 +155,66 @@ public static Message createIotHubMessageUtf8(@NonNull String telemetryName, @No
-            put("ac", ackCode);
-            put("av", ackVersion);
-        }};
+        return createWritablePropertyResponse(new HashMap<String, Object>()
+        {{
+            put(propertyName, propertyValue);
+        }}, ackCode, ackVersion, ackDescription);
+    }
 
-        if (ackDescription != null && !ackDescription.isEmpty()) {
-            embeddedProperty.put("ad", ackDescription);
-        }
+    /**
+     * Creates a response to a write request on a device property.
+     * @param propertyPairs The name and unserialized value of the property to report.
+     * @param ackCode The acknowledgment code from the device, for the embedded value property update.
+     * @param ackVersion The version no. of the service-initiated read-write property update.
+     * @param ackDescription (optional) The description from the device, accompanying the embedded value property update.
+     * @return The property patch for embedded value property updates for read-write properties.
+     *
+     * The property patch is created in the below format:
+     * {
+     *     "samplePropertyName": {
+     *         "value": 20,
+     *         "ac": 200,
+     *         "av": 5,
+     *         "ad": "The update was successful."
+     *     }
+     * }
+     */
+    public static Set<Property> createWritablePropertyResponse(
+            @NonNull HashMap<String, Object> propertyPairs,
+            @NonNull final Integer ackCode,
+            @NonNull final Long ackVersion,
+            String ackDescription) {
+
+        Set<Property> response = new HashSet<>();
+
+        propertyPairs.forEach((k,v) ->
+            {
+                if(k == null || k.isEmpty())
+                {
+                    throw new IllegalArgumentException("One of the propertyPairs keys was null, empty, or white space.");
+                }
 
-        return singleton(new Property(propertyName, embeddedProperty));
+                if (ackDescription != null && !ackDescription.isEmpty()) {
+                    response.add(new Property(k, new WritablePropertyResponse(v, ackCode, ackVersion)));
+                }
+                else
+                {
+                    response.add(new Property(k, new WritablePropertyResponse(v, ackCode, ackVersion, ackDescription)));
+                }
+            });
+
+        return response;
     }
 
     /**
-     * Create a key-embedded value property patch for read-write properties.
-     * Embedded value property updates are sent from a device in response to a service-initiated read-write property update.
+     * Creates a response to a write request on a device property.
      * @param propertyName The property name, as defined in the DTDL interface.
      * @param propertyValue The property value, in the format defined in the DTDL interface.
      * @param componentName (optional) The name of the component in which the property is defined. Can be null for property defined under the root interface.
```

Comment:
is componentName still considered optional since you added the NonNull?

Reply:
will address this